### PR TITLE
Build CLI only for host platform

### LINKED ISSUE #546: bin/docker-build should only build the CLI executable for the host platform by default
The objective here is to speed up and simplify the edit-build-test cycle by removing inefficiency in how the CLI executable is built, while still being able to reply on `bin/docker-build` to ensure that the CLI executable is built with the same version tag as the rest of Conduit.

Currently:

1. `bin/docker-build` executes `bin/docker-build-cli` (and other things, not relevant to this issue).
2. `bin/docker-build-cli` executes `bin/docker-build-cli-bin`
3. `bin/docker-build-cli` then does the following things:
    1. 3 executables, one for each target platform, are built in a Docker container:
        ```
        RUN CGO_ENABLED=0 GOOS=darwin  go build -o /out/conduit-darwin [...]
        RUN CGO_ENABLED=0 GOOS=linux   go build -o /out/conduit-linux [...]
        RUN CGO_ENABLED=0 GOOS=windows go build -o /out/conduit-windows [...]
        ```
    2. A new Docker container is built, containing just those three output files.
    3. All three executables are extracted out of the Docker image created in step 3.2 into the local filesystem at `target/cli/${os}/conduit`.

In typical usage of `bin/docker-build`, we have almost zero use for anything except the CLI executable for the host platform. That is, if I'm running `bin/docker-build` on Linux in my edit-build-test cycle then I don't care about any of the docker images created above; as far as the `conduit` executable is concerned, I only care that `conduit` is built and written to `target/cli/linux/conduit`. In particular, I never need the conduit images to be built for Windows or Mac. Similarly, somebody running on a Mac doesn't usually need the images built for Linux or Windows.

We should redo this so that the default action of bin/docker-build is to skip as much of the above as possible and directly build `target/cli/${os}/conduit` where `${os}` is the host OS, and only build the other things (the docker images and the executables for other platforms) if/when specifically requested.

### PR DESCRIPTION
**Pull request by alenkacz**
The dockerfile still by default builds all platforms, the build script is working only for a target one.

If that is not enough, I guess the script can also accept some kind of argument and be used optionally to build all platforms at once.

Fixes #546

Signed-off-by: Alena Varkockova <varkockova.a@gmail.com>

### CODE REVIEW

File: bin/docker-build-cli-bin [hunk 1]
```diff
@@ -22,22 +22,28 @@ validate_go_deps_tag $dockerfile
     $bindir/docker-build-go-deps
 ) >/dev/null
 
+CURRENT_PLATFORM=$(uname)
```
Comment (briansmith) [inline bin/docker-build-cli-bin]: We do need a way to build the Docker container the way it is now, with all three platforms included in the container, by executing bin/docker-build. The reason is that the official release process, which isn't documented in this open source repo yet (I hope to do so soon) consists of `bin/docker-build` followed by uploading all three platforms' `conduit` CLI apps to GitHub (amongst other steps).

Also, more generally, it might be confusing and/or error-prone to build the CLI docker image in a way that's different from its normal intended configuration. And also, if we can avoid Docker in the first place, it would be faster yet. On my laptop, as much of the slowness of building the CLI comes from Docker and virtual machine overhead as it does from the actual `go build` steps in the Dockerfile.

I suggest that we do this a bit differently:
* Create a `bin/fast-build` script
* Inside `bin/fast-build`, execute something like `CONDUIT_SKIP_CLI_CONTAINER=1 bin/docker-build`
* In `bin/docker-build`, skip `docker-build-cli-bin` completely if `$CONDUIT_SKIP_CLI_CONTAINER` is set.
* In `bin/fast-build`, add a step that does `CGO_ENABLED=0 bin/go-build  go build -installsuffix cgo -o target/cli/${OS}/conduit -ldflags "${GO_LDFLAGS}" where `GO_LDFLAGS` is set like it is in the Dockerfile.
* (Perhaps as a follow-up) Find some convenient way to centralize the command line flags for `go build` and `go install` in one file that is used by all the Dockerfiles and the fast-build script.

This way, we would create a fast-path that eliminates all the unnecessary overhead, both Docker and the waste of building for usually-irrelevant platforms, with minimal risk to the production build system.

WDYT?

**Comment by alenkacz** [comment]
Sounds good, let me think about it when I come back from vacation on Apr 8 :) 🏝

File: bin/docker-build-cli-bin (hunk 1 above)
Comment (alenkacz) [reply bin/docker-build-cli-bin]: is the docker really such a bottleneck? my only problem with the thing that you are suggesting is that we will introduce a second build script that is not run in CI... so it is potentially one more thing that have to be maintained in the future and if some breaking change in a way the CLI is built is introduced, this new script might get broken and we won't even know that from CI...

I would personally stick with one way to build this but if you say that the performance gain is actually worth it, I'll implement your way :)

let me know...

Comment (alenkacz) [reply bin/docker-build-cli-bin]: @briansmith ping? :)

Comment (briansmith) [reply bin/docker-build-cli-bin]: OK. I think it is fine to limit the scope of this PR to more what you originally submitted. Below I suggest smaller changes that are necessary for it to be merged.

File: bin/docker-build-cli-bin [hunk 2]
```diff
@@ -22,22 +22,28 @@ validate_go_deps_tag $dockerfile
...
+target_platform="windows"
```
Comment (briansmith) [inline bin/docker-build-cli-bin]: `target_platform` should be called `host_platform`.

File: cli/Dockerfile-bin [hunk 3]
```diff
@@ -1,15 +1,16 @@
 ## compile binaries
 FROM gcr.io/runconduit/go-deps:5a44df7f as golang
 ARG CONDUIT_VERSION
+ARG PLATFORMS="darwin linux windows"
```
Comment (briansmith) [inline cli/Dockerfile-bin]: Did you intend to change `bin/docker-build-cli-bin` so that we can override `PLATFORMS` so that only one platform is built? Otherwise, I don't understand why the changes to this file are useful as-is.

Comment (alenkacz) [reply cli/Dockerfile-bin]: oh thanks, I must have removed this by mistake :(

**Comment by alenkacz** [comment]
@briansmith ready to be re-reviewed. 

If you still think the fast-build you proposed makes sense - let me know and I can do that too :-)

**Comment by briansmith** [comment]
> @briansmith ready to be re-reviewed.

I think you might have overlooked one requirement:

Please update the documentation in BUILD.md to indicate how one would run bin/docker-build for two scenerios:

* How to build and extract all three platforms' conduit executables.
* How to build and extract only the host platforms' conduit executable.

In particular, looking at the PR, it looks like it's not possible to build the cli-bin container such that it contains all three platforms' binaries anymore. However, we need to be able to do that when doing Conduit releases. Currently our instructions look like this:

> DOCKER_TRACE=1 bin/docker-build
> mv target/cli/darwin/conduit conduit-${TAG:1}-darwin
> mv target/cli/linux/conduit conduit-${TAG:1}-linux
> mv target/cli/windows/conduit conduit-${TAG:1}-windows.exe

Also, if/when we build a docker image that contains only one platform instead of all three, then we shouldn't give that docker image the same name that we'd give the one that contains all three. Otherwise it will be very confusing when you have a docker image named `cli-bin` because you won't know what it contains. I suggest that we call the image `cli-bin-$platform` where `$platform` is the platform for which the CLI image was built.

> If you still think the fast-build you proposed makes sense - let me know and I can do that too :-)

Yes, I am very interested in the fast-build idea. We should either finish this PR up first (fix up and merge, or close without merging).

**Comment by alenkacz** [comment]
@briansmith if you really don't find the approach in this PR useful I am fine with not investing any more time and I'll go the fast build way. That's why I was asking for your input :)

**Comment by briansmith** [comment]
> @briansmith if you really don't find the approach in this PR useful I am fine with not investing any more time (and close this) and I'll go the fast build way. That's why I was asking for your input :)

I think this PR is a clear improvement if we add the instructions to BUILD.md for how to build all three platforms, so I'd be happy to merge it when it is ready. However, if you want to skip to the fast-build thing then that's fine too.

**Comment by alenkacz** [comment]
just to post some update here, I decided to go the fast-build way in the end, because I think it actually makes sense (and I should have listened to you in the first place) :)

but I am struggling there with something so it takes time :)
